envoy/http/header_validator.h

```diff
@@ -0,0 +1,81 @@
+#pragma once
+
+#include "envoy/config/typed_config.h"
+#include "envoy/http/header_map.h"
+#include "envoy/server/factory_context.h"
+
+namespace Envoy {
+namespace Http {
+
+/**
+ * Interface for header key formatters that are stateful. A formatter is created during decoding
+ * headers, attached to the header map, and can then be used during encoding for reverse
+ * translations if applicable.
+ */
+class HeaderValidator {
+public:
+  virtual ~HeaderValidator() = default;
+
+  enum class HeaderEntryValidationResult { Accept, Reject };
+
+  /**
+   * Method for validating a header entry. This is called for both request and response headers.
```

Comment:
I assume this only does generic validation, not specific things like content and host header?  otherwise I think we need directionality (request/response)

Reply:
Good point. I'm not completely sure if the generic validation would be enough. I think H/2 codec validates allowed request/response header as it parses HEADERS frame. I've added directionality just in case. We can tune this later if need be.